app/js/lib/foxbox/network.js

```diff
@@ -172,49 +175,70 @@ export default class Network {
   }
 
   /**
-   * Ping the box to detect whether we connect locally or remotely.
+   * Ping the box to detect whether we connect locally or remotely. Since
+   * 'online' state depends on two factors: local and remote server
+   * availability, there is a slight chance that this method will cause two
+   * events e.g. if previously box was available only locally and now it's
+   * available only remotely we'll likely generate event indicating that box
+   * went offline following by another event indicating that box is online
+   * again.
    *
    * @private
    */
   [p.pingBox]() {
-    const promises = [];
-    promises.push(this[p.ping](`${this.localOrigin}/ping`)
-      .then(() => {
-        this[p.local] = true;
-      })
-      .catch(() => {
-        this[p.local] = false;
-      }));
+    const previousState = this[p.local] || this[p.remote];
+
+    this[p.ping](`${this.localOrigin}/ping`)
+      .then((isOnline) => this[p.onPong](previousState, p.local, isOnline));
 
     // @todo Find a better way to detect if a tunnel connection is active.
     if (this[p.settings].tunnelOrigin) {
-      promises.push(this[p.ping](`${this.tunnelOrigin}/ping`)
-        .then(() => {
-          this[p.remote] = true;
-        })
-        .catch(() => {
-          this[p.remote] = false;
-        }));
+      this[p.ping](`${this.tunnelOrigin}/ping`)
+        .then((isOnline) => this[p.onPong](previousState, p.remote, isOnline));
     }
-
-    Promise.all(promises).then(() => {
-      this[p.pingCallback](this[p.local] || this[p.remote]);
-    });
   }
 
   /**
-   * Get a URL but don't process the response.
+   * Performs HTTP 'GET' request to the specified URL. Returns 'true' if
+   * response was successful (any of 200-299 status codes) or 'false'
+   * otherwise.
    *
    * @param {string} url The URL to send the request to.
-   * @return {Promise}
+   * @return {Promise<boolean>}
    * @private
    */
   [p.ping](url) {
     return fetch(url, { cache: 'no-store' })
       .then((res) => res.ok)
       .catch((error) => {
-        console.error('Error occurred while pinging content: ', error);
-        throw error;
+        console.error('Error occurred while pinging content: %o', error);
+        return false;
       });
   }
+
+  /**
+   * Process ping response (pong). If 'online' state is changed we emit
+   * 'online-state-changed' event.
+   *
+   * @param {boolean} previousOnlineState Previous 'online' state.
+   * @param {Symbol} localOrRemote Symbol indicating whether we process local
+   * pong or remote one.
+   * @param {boolean} isOnline Flag that indicates whether pinged server
+   * successfully responded to ping request.
+   * @private
+   */
+  [p.onPong](previousOnlineState, localOrRemote, isOnline) {
```

Comment:
I don't really happy about this method, alternative was to create function inside ```pingBox``` on every ping. And I don't like alternative even more...